### ISSUE TITLE
!= false filter doesn't work in MySQL

### ISSUE THREAD

**Issue by cobbspur** [metabase/metabase#1680]
Hi,

I was running a simple query testing for boolean values. My data values are either true/false.
 
Everything is fine when using 'Is', but if you test 'Is Not false' you get a query error.

**Comment by tlrobinson** [comment]
@cobbspur What is your database and column type? If it's a boolean column you shouldn't even be getting an "Is not" operator option, since "Is not false" is equivalent to "Is true" (although I suppose that's not technically true if you want to match `true` or `NULL`). Does the operator pane look like this?

![screenshot 2015-12-14 13 37 22](https://cloud.githubusercontent.com/assets/18193/11794857/57a72562-a268-11e5-90dd-8abb7ab83b1c.png)

**Comment by cobbspur** [comment]
Hi, 

The database is MySQL and actually my column type is 1/0 not true/false - sorry

![screen shot 2015-12-14 at 22 00 24](https://cloud.githubusercontent.com/assets/4405395/11795379/326e0f40-a2ae-11e5-98e4-db508fe83d49.png)

Don't get me wrong, I know I can do 'is true' and 'is false' and this is not stopping me from doing what I want, but 'is not false' should be the same as 'is true' in boolean logic so it seems to be a bug

**Comment by camsaul** [comment]
That's true of boolean logic, but [we're dealing with ternary logic](https://en.wikipedia.org/wiki/Three-valued_logic#Application_in_SQL) here. e.g.

```
value | true? (= 1) | false? (= 0) | not true? (<> 1) | not false? (<> 0) |
------+-------------+--------------+------------------+-------------------+
    0 |          no |         yes  |              yes |                no |
    1 |         yes |          no  |               no |               yes |
 NULL |          no |          no  |              yes |               yes |
```

**Comment by cobbspur** [comment]
@camsaul The reason I ran into this is that it is quicker to to click 'is not' than deselect false and select true. If the UI available was the same as @lrobinson showed then this would not be an issue.

**Comment by cobbspur** [comment]
The query being built is 

    SQL: ?
    SELECT ...
    FROM `users`
    WHERE ((`users`.`seen_welcome_message` IS NOT ?))
    LIMIT 2000

The log from metabase is:

    You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '0)) LIMIT 2000' at line 1

**Comment by camsaul** [comment]
Thanks for the clarification. That definitely looks like a backend bug. We're generating a `IS NOT 0`, which isn't valid MySQL syntax. `<> 0`, `!= 0`, and `IS NOT FALSE` are all legal and mean the same thing . I'll look into this 😳

**Comment by tlrobinson** [comment]
@camsaul Separately, should we be treating `tinyint(1)` or `bit` columns as booleans?

**Comment by agilliland** [comment]
I believe this is a good example of why physical column types are not meant to be 1:1 with the data type meant for user interaction.  We are slowly building up examples of this, but basically we want to allow smart casting so that we can map the physical column types to something more sensible for users, such as a 1 or 0 -> true/false.

I agree with Tom with respect to the fact that this is probably something that should be editable by the user because there are cases where tinyint(1) may not actually be intended as a boolean.

**Comment by camsaul** [comment]
@tlrobinson It depends on the DB. MySQL for example doesn't *actually* have a native boolean type so the recommendation is just to use `tinyint(1)` or `bit`.`BOOLEAN` is an alias of `TINYINT(1)`, `TRUE` is an alias for `1`, and `FALSE` is an alias for `0`. As much as possible, I've tried to write all the drivers in a way where they each handle their own idiosyncrasies, so the frontend doesn't need to know or worry about them.

`BIT` fields are already treated as booleans automatically in our H2 and in SQL Server drivers, which has no native boolean type either. I think I wrote the MySQL driver long enough ago that I hadn't yet fleshed out the pattern for handling this case, so it doesn't have the extra bit of magic yet.

At any rate, the MySQL query should have worked either way, if it weren't for the bug in the Query Processor

**Comment by camsaul** [comment]
@agilliland We *don't* have a "1:1" mapping between physical column types and what the user sees. In this case the column would be showing the boolean widget like in Tom's screenshot if it weren't for the bug in the MySQL driver.

Also, I don't think Tom said anywhere that this is something that he thinks should be editable by the user? And in what use cases do you see people wanting to use `TINYINT(1)` as something besides a boolean? The only possible option is making it an `:IntegerField` with two possible values -- either `0` or `1`.

**Comment by agilliland** [comment]
The `tinyint(1)` column type in mysql is not limited to a 0 or 1 value, it simply represents a 1 byte integer.  http://dev.mysql.com/doc/refman/5.7/en/integer-types.html

```
mysql> describe foo;
+-------+------------+------+-----+---------+-------+
| Field | Type       | Null | Key | Default | Extra |
+-------+------------+------+-----+---------+-------+
| t     | tinyint(1) | YES  |     | NULL    |       |
+-------+------------+------+-----+---------+-------+
1 row in set (0.01 sec)

mysql> select * from foo;
+------+
| t    |
+------+
|    5 |
|  126 |
+------+
2 rows in set (0.00 sec)
```

Tom didn't say anything about this being editable by the user, I did.  This is the same reason why we allow long integer values to optionally be treated as timestamps, it's because there are cases where the physical storage of a value does not seamlessly indicate the intended meaning behind the column.  

So all I am pointing out here is that this is an example where seeing a column type of tinyint(1) is ambiguous and thus we should provide some way to resolve that ambiguity by allowing the user to tweak the metadata and make it clear what was intended.

**Comment by camsaul** [comment]
@agilliland Yes, `tinyint` is a 1-byte integer. But we're talking about `tinyint(1)`, which is *not* treated the same way as plain `tinyint`. [`tinyint(1)` is explicitly treated as a boolean by MySQL](http://dev.mysql.com/doc/refman/5.0/en/numeric-type-overview.html):

```
A value of zero is considered false. Nonzero values are considered true.
```

Here's some sample code if you want to test it for yourself. The official MySQL JDBC driver returns the columns as `true`/`false`, regardless of their actual value. So to treat them as anything else would actually mean we'd have to do extra casting.
 
```clojure
(def fancy-birds
  (i/create-database-definition "boolean_test"
    ["fancy_birds" [{:field-name "name"
                     :base-type  :CharField}
                    {:field-name "fancy"
                     :base-type  {:native "TINYINT(1)"}}]
     [["Toucan" true]
      ["Pigeon" false]
      ["Penguin" true]
      ["Crow" false]
      ["Eagle" true]
      ["Parrot" true]
      ["Chicken" false]
      ["Hawk" 2]
      ["Sea Gull" 4]
      ["Pelican" 8]
      ["Bluejay" 16]
      ["Sparrow" 32]
      ["Turkey" 64]]]))

(datasets/with-engine :mysql
  (data/with-db (data/get-or-create-database! fancy-birds)
    (Q aggregate rows of fancy_birds return rows)))

;; ->

[[1 true "Toucan"]
 [2 false "Pigeon"]
 [3 true "Penguin"]
 [4 false "Crow"]
 [5 true "Eagle"]
 [6 true "Parrot"]
 [7 false "Chicken"]
 [8 true "Hawk"]
 [9 true "Sea Gull"]
 [10 true "Pelican"]
 [11 true "Bluejay"]
 [12 true "Sparrow"]
 [13 true "Turkey"]]
```

**Comment by camsaul** [comment]
Fixed by #1686
